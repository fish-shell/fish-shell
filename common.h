/** \file common.h
	Prototypes for various functions, mostly string utilities, that are used by most parts of fish.
*/

#ifndef FISH_COMMON_H
/**
   Header guard
*/
#define FISH_COMMON_H

#include <stdlib.h>
#include <stdio.h>
#include <wchar.h>
#include <termios.h>
#include <string>
#include <sstream>
#include <vector>
#include <pthread.h>
#include <string.h>

#include <errno.h>
#include <assert.h>
#include "util.h"

class completion_t;

/* Common string type */
typedef std::wstring wcstring;
typedef std::vector<wcstring> wcstring_list_t;

/**
   Maximum number of bytes used by a single utf-8 character
*/
#define MAX_UTF8_BYTES 6

/**
   This is in the unicode private use area.
*/
#define ENCODE_DIRECT_BASE 0xf100

/**
  Highest legal ascii value
*/
#define ASCII_MAX 127u

/**
  Highest legal 16-bit unicode value
*/
#define UCS2_MAX 0xffffu

/**
  Highest legal byte value
*/
#define BYTE_MAX 0xffu

/**
  Escape special fish syntax characters like the semicolon
 */
#define UNESCAPE_SPECIAL 1

/**
  Allow incomplete escape sequences
 */
#define UNESCAPE_INCOMPLETE 2

/* Flags for the escape() and escape_string() functions */
enum
{
    /** Escape all characters, including magic characters like the semicolon */
    ESCAPE_ALL = 1 << 0,

    /** Do not try to use 'simplified' quoted escapes, and do not use empty quotes as the empty string */
    ESCAPE_NO_QUOTED = 1 << 1,

    /** Do not escape tildes */
    ESCAPE_NO_TILDE = 1 << 2
};
typedef unsigned int escape_flags_t;

/**
 Helper macro for errors
 */
#define VOMIT_ON_FAILURE(a) do { if (0 != (a)) { int err = errno; fprintf(stderr, "%s failed on line %d in file %s: %d (%s)\n", #a, __LINE__, __FILE__, err, strerror(err)); abort(); }} while (0)

/** Exits without invoking destructors (via _exit), useful for code after fork. */
void exit_without_destructors(int code) __attribute__((noreturn));

/**
	Save the shell mode on startup so we can restore them on exit
*/
extern struct termios shell_modes;

/**
   The character to use where the text has been truncated. Is an
   ellipsis on unicode system and a $ on other systems.
*/
extern wchar_t ellipsis_char;

/* Character representing an omitted newline at the end of text */
extern wchar_t omitted_newline_char;

/**
   The verbosity level of fish. If a call to debug has a severity
   level higher than \c debug_level, it will not be printed.
*/
extern int debug_level;

/**
   Profiling flag. True if commands should be profiled.
*/
extern char *profile;

/**
   Name of the current program. Should be set at startup. Used by the
   debug function.
*/
extern const wchar_t *program_name;

/**
   This macro is used to check that an input argument is not null. It
   is a bit lika a non-fatal form of assert. Instead of exit-ing on
   failiure, the current function is ended at once. The second
   parameter is the return value of the current function on failiure.
*/
#define CHECK( arg, retval )											\
	if (!(arg)) 														\
	{																	\
		debug( 0,														\
			   "function %s called with null value for argument %s. ",  \
			   __func__,												\
			   #arg );													\
		bugreport();													\
		show_stackframe();												\
		return retval;													\
	}

/**
   Pause for input, then exit the program. If supported, print a backtrace first.
*/
#define FATAL_EXIT()											\
	{															\
		char exit_read_buff;			\
		show_stackframe();										\
		read( 0, &exit_read_buff, 1 );			\
		exit_without_destructors( 1 );												\
	}															\
	

/**
   Exit program at once, leaving an error message about running out of memory.
*/
#define DIE_MEM()														\
	{																	\
		fwprintf( stderr,												\
				  L"fish: Out of memory on line %ld of file %s, shutting down fish\n", \
				  (long)__LINE__,												\
				  __FILE__ );											\
		FATAL_EXIT();														\
	}

/**
   Check if signals are blocked. If so, print an error message and
   return from the function performing this check.
*/
#define CHECK_BLOCK(retval) 											\
	if (signal_is_blocked()) 											\
	{																	\
		debug( 0,														\
			    "function %s called while blocking signals. ",  		\
			   __func__);												\
		bugreport();													\
		show_stackframe();												\
		return retval;													\
	}

/**
   Shorthand for wgettext call
*/
#define _(wstr) wgettext((const wchar_t *)wstr)

/**
   Noop, used to tell xgettext that a string should be translated,
   even though it is not directly sent to wgettext.
*/
#define N_(wstr) wstr

/**
   Check if the specified string element is a part of the specified string list
 */
#define contains( str,... ) contains_internal( str, __VA_ARGS__, NULL )

/**
  Print a stack trace to stderr
*/
void show_stackframe();


/**
   Read a line from the stream f into the string. Returns
   the number of bytes read or -1 on failiure.

   If the carriage return character is encountered, it is
   ignored. fgetws() considers the line to end if reading the file
   results in either a newline (L'\n') character, the null (L'\\0')
   character or the end of file (WEOF) character.
*/
int fgetws2(wcstring *s, FILE *f);


/**
 Returns a  wide character string equivalent of the
 specified multibyte character string

 This function encodes illegal character sequences in a reversible
 way using the private use area.
 */
wcstring str2wcstring(const char *in);
wcstring str2wcstring(const char *in, size_t len);
wcstring str2wcstring(const std::string &in);

/**
   Returns a newly allocated multibyte character string equivalent of
   the specified wide character string

   This function decodes illegal character sequences in a reversible
   way using the private use area.
*/
char *wcs2str(const wchar_t *in);
std::string wcs2string(const wcstring &input);

/** Test if a string prefixes another. Returns true if a is a prefix of b */
bool string_prefixes_string(const wcstring &proposed_prefix, const wcstring &value);
bool string_prefixes_string(const wchar_t *proposed_prefix, const wcstring &value);

/** Test if a string is a suffix of another */
bool string_suffixes_string(const wcstring &proposed_suffix, const wcstring &value);
bool string_suffixes_string(const wchar_t *proposed_suffix, const wcstring &value);


/** Test if a string prefixes another without regard to case. Returns true if a is a prefix of b */
bool string_prefixes_string_case_insensitive(const wcstring &proposed_prefix, const wcstring &value);

/** Test if a list contains a string using a linear search. */
bool list_contains_string(const wcstring_list_t &list, const wcstring &str);


void assert_is_main_thread(const char *who);
#define ASSERT_IS_MAIN_THREAD_TRAMPOLINE(x) assert_is_main_thread(x)
#define ASSERT_IS_MAIN_THREAD() ASSERT_IS_MAIN_THREAD_TRAMPOLINE(__FUNCTION__)

void assert_is_background_thread(const char *who);
#define ASSERT_IS_BACKGROUND_THREAD_TRAMPOLINE(x) assert_is_background_thread(x)
#define ASSERT_IS_BACKGROUND_THREAD() ASSERT_IS_BACKGROUND_THREAD_TRAMPOLINE(__FUNCTION__)

/* Useful macro for asserting that a lock is locked. This doesn't check whether this thread locked it, which it would be nice if it did, but here it is anyways. */
void assert_is_locked(void *mutex, const char *who, const char *caller);
#define ASSERT_IS_LOCKED(x) assert_is_locked((void *)(&x), #x, __FUNCTION__)

/** Format the specified size (in bytes, kilobytes, etc.) into the specified stringbuffer. */
wcstring format_size(long long sz);

/** Version of format_size that does not allocate memory. */
void format_size_safe(char buff[128], unsigned long long sz);

/** Our crappier versions of debug which is guaranteed to not allocate any memory, or do anything other than call write(). This is useful after a call to fork() with threads. */
void debug_safe(int level, const char *msg, const char *param1 = NULL, const char *param2 = NULL, const char *param3 = NULL, const char *param4 = NULL, const char *param5 = NULL, const char *param6 = NULL, const char *param7 = NULL, const char *param8 = NULL, const char *param9 = NULL, const char *param10 = NULL, const char *param11 = NULL, const char *param12 = NULL);

/** Writes out a long safely */
void format_long_safe(char buff[128], long val);
void format_long_safe(wchar_t buff[128], long val);


template<typename T>
T from_string(const wcstring &x)
{
    T result;
    std::wstringstream stream(x);
    stream >> result;
    return result;
}

template<typename T>
T from_string(const std::string &x)
{
    T result = T();
    std::stringstream stream(x);
    stream >> result;
    return result;
}

template<typename T>
wcstring to_string(const T &x)
{
    std::wstringstream stream;
    stream << x;
    return stream.str();
}

/* wstringstream is a huge memory pig. Let's provide some specializations where we can. */
template<>
inline wcstring to_string(const long &x)
{
    wchar_t buff[128];
    format_long_safe(buff, x);
    return wcstring(buff);
}

template<>
inline bool from_string(const std::string &x)
{
    return ! x.empty() && strchr("YTyt1", x.at(0));
}

template<>
inline bool from_string(const wcstring &x)
{
    return ! x.empty() && wcschr(L"YTyt1", x.at(0));
}

template<>
inline wcstring to_string(const int &x)
{
    return to_string(static_cast<long>(x));
}


/* Helper class for managing a null-terminated array of null-terminated strings (of some char type) */
template <typename CharType_t>
class null_terminated_array_t
{
    CharType_t **array;

    typedef std::basic_string<CharType_t> string_t;
    typedef std::vector<string_t> string_list_t;

    void swap(null_terminated_array_t<CharType_t> &him)
    {
        std::swap(array, him.array);
    }

    /* Silly function to get the length of a null terminated array of...something */
    template <typename T>
    static size_t count_not_null(const T *arr)
    {
        size_t len;
        for (len=0; arr[len] != T(0); len++)
            ;
        return len;
    }

    size_t size() const
    {
        return count_not_null(array);
    }

    void free(void)
    {
        if (array != NULL)
        {
            for (size_t i = 0; array[i] != NULL; i++)
            {
                delete [] array[i];
            }
            delete [] array;
            array = NULL;
        }
    }

public:
    null_terminated_array_t() : array(NULL) { }
    null_terminated_array_t(const string_list_t &argv) : array(NULL)
    {
        this->set(argv);
    }
    ~null_terminated_array_t()
    {
        this->free();
    }

    /** operator=. Notice the pass-by-value parameter. */
    null_terminated_array_t& operator=(null_terminated_array_t rhs)
    {
        if (this != &rhs)
            this->swap(rhs);
        return *this;
    }

    /* Copy constructor. */
    null_terminated_array_t(const null_terminated_array_t &him) : array(NULL)
    {
        this->set(him.array);
    }

    void set(const string_list_t &argv)
    {
        /* Get rid of the old argv */
        this->free();

        /* Allocate our null-terminated array of null-terminated strings */
        size_t i, count = argv.size();
        this->array = new CharType_t * [count + 1];
        for (i=0; i < count; i++)
        {
            const string_t &str = argv.at(i);
            this->array[i] = new CharType_t [1 + str.size()];
            std::copy(str.begin(), str.end(), this->array[i]);
            this->array[i][str.size()] = CharType_t(0);
        }
        this->array[count] = NULL;
    }

    void set(const CharType_t * const *new_array)
    {
        if (new_array == array)
            return;

        /* Get rid of the old argv */
        this->free();

        /* Copy the new one */
        if (new_array)
        {
            size_t i, count = count_not_null(new_array);
            this->array = new CharType_t * [count + 1];
            for (i=0; i < count; i++)
            {
                size_t len = count_not_null(new_array[i]);
                this->array[i] = new CharType_t [1 + len];
                std::copy(new_array[i], new_array[i] + len, this->array[i]);
                this->array[i][len] = CharType_t(0);
            }
            this->array[count] = NULL;
        }
    }

    CharType_t **get()
    {
        return array;
    }
    const CharType_t * const *get() const
    {
        return array;
    }

    string_list_t to_list() const
    {
        string_list_t lst;
        if (array != NULL)
        {
            size_t count = this->size();
            lst.reserve(count);
            lst.insert(lst.end(), array, array + count);
        }
        return lst;
    }
};

/* Helper function to convert from a null_terminated_array_t<wchar_t> to a null_terminated_array_t<char_t> */
null_terminated_array_t<char> convert_wide_array_to_narrow(const null_terminated_array_t<wchar_t> &arr);

/* Helper class to cache a narrow version of a wcstring in a malloc'd buffer, so that we can read it after fork() */
class narrow_string_rep_t
{
private:
    const char *str;

    /* No copying */
    narrow_string_rep_t &operator=(const narrow_string_rep_t &);
    narrow_string_rep_t(const narrow_string_rep_t &x);

public:
    ~narrow_string_rep_t()
    {
        free((void *)str);
    }

    narrow_string_rep_t() : str(NULL) {}

    void set(const wcstring &s)
    {
        free((void *)str);
        str = wcs2str(s.c_str());
    }

    const char *get() const
    {
        return str;
    }
};

bool is_forked_child();

/* Basic scoped lock class */
class scoped_lock
{
    pthread_mutex_t *lock_obj;
    bool locked;

    /* No copying */
    scoped_lock &operator=(const scoped_lock &);
    scoped_lock(const scoped_lock &);

public:
    void lock(void);
    void unlock(void);
    scoped_lock(pthread_mutex_t &mutex);
    ~scoped_lock();
};

/* Wrapper around wcstok */
class wcstokenizer
{
    wchar_t *buffer, *str, *state;
    const wcstring sep;

    /* No copying */
    wcstokenizer &operator=(const wcstokenizer &);
    wcstokenizer(const wcstokenizer &);

public:
    wcstokenizer(const wcstring &s, const wcstring &separator);
    bool next(wcstring &result);
    ~wcstokenizer();
};

/**
 Appends a path component, with a / if necessary
 */
void append_path_component(wcstring &path, const wcstring &component);

wcstring format_string(const wchar_t *format, ...);
wcstring vformat_string(const wchar_t *format, va_list va_orig);
void append_format(wcstring &str, const wchar_t *format, ...);

/**
   Returns a newly allocated wide character string array equivalent of
   the specified multibyte character string array
*/
char **wcsv2strv(const wchar_t * const *in);

/**
   Test if the given string is a valid variable name.

   \return null if this is a valid name, and a pointer to the first invalid character otherwise
*/

wchar_t *wcsvarname(const wchar_t *str);


/**
   Test if the given string is a valid function name.

   \return null if this is a valid name, and a pointer to the first invalid character otherwise
*/

const wchar_t *wcsfuncname(const wchar_t *str);

/**
   Test if the given string is valid in a variable name

   \return 1 if this is a valid name, 0 otherwise
*/

int wcsvarchr(wchar_t chr);


/**
   A wcswidth workalike. Fish uses this since the regular wcswidth seems flaky.
*/
int my_wcswidth(const wchar_t *c);

/**
   This functions returns the end of the quoted substring beginning at
   \c in. The type of quoting character is detemrined by examining \c
   in. Returns 0 on error.

   \param in the position of the opening quote
*/
wchar_t *quote_end(const wchar_t *in);

/**
   A call to this function will reset the error counter. Some
   functions print out non-critical error messages. These should check
   the error_count before, and skip printing the message if
   MAX_ERROR_COUNT messages have been printed. The error_reset()
   should be called after each interactive command executes, to allow
   new messages to be printed.
*/
void error_reset();

/**
   This function behaves exactly like a wide character equivalent of
   the C function setlocale, except that it will also try to detect if
   the user is using a Unicode character set, and if so, use the
   unicode ellipsis character as ellipsis, instead of '$'.
*/
wcstring wsetlocale(int category, const wchar_t *locale);

/**
   Checks if \c needle is included in the list of strings specified. A warning is printed if needle is zero.

   \param needle the string to search for in the list

   \return zero if needle is not found, of if needle is null, non-zero otherwise
*/
__sentinel bool contains_internal(const wchar_t *needle, ...);
__sentinel bool contains_internal(const wcstring &needle, ...);

/**
   Call read while blocking the SIGCHLD signal. Should only be called
   if you _know_ there is data available for reading, or the program
   will hang until there is data.
*/
long read_blocked(int fd, void *buf, size_t count);

/**
   Loop a write request while failiure is non-critical. Return -1 and set errno
   in case of critical error.
 */
ssize_t write_loop(int fd, const char *buff, size_t count);

/**
   Loop a read request while failiure is non-critical. Return -1 and set errno
   in case of critical error.
 */
ssize_t read_loop(int fd, void *buff, size_t count);


/**
   Issue a debug message with printf-style string formating and
   automatic line breaking. The string will begin with the string \c
   program_name, followed by a colon and a whitespace.

   Because debug is often called to tell the user about an error,
   before using wperror to give a specific error message, debug will
   never ever modify the value of errno.

   \param level the priority of the message. Lower number means higher priority. Messages with a priority_number higher than \c debug_level will be ignored..
   \param msg the message format string.

   Example:

   <code>debug( 1, L"Pi = %.3f", M_PI );</code>

   will print the string 'fish: Pi = 3.141', given that debug_level is 1 or higher, and that program_name is 'fish'.
*/
void debug(int level, const char *msg, ...);
void debug(int level, const wchar_t *msg, ...);

/**
   Replace special characters with backslash escape sequences. Newline is
   replaced with \n, etc.

   \param in The string to be escaped
   \param escape_all Whether all characters wich hold special meaning in fish (Pipe, semicolon, etc,) should be escaped, or only unprintable characters
   \return The escaped string, or 0 if there is not enough memory
*/

wchar_t *escape(const wchar_t *in, escape_flags_t flags);
wcstring escape_string(const wcstring &in, escape_flags_t flags);

/**
   Expand backslashed escapes and substitute them with their unescaped
   counterparts. Also optionally change the wildcards, the tilde
   character and a few more into constants which are defined in a
   private use area of Unicode. This assumes wchar_t is a unicode
   character set.

   The result must be free()d. The original string is not modified. If
   an invalid sequence is specified, 0 is returned.

*/
wchar_t *unescape(const wchar_t * in,
                  int escape_special);

bool unescape_string(wcstring &str,
                     int escape_special);


/**
    Returns the width of the terminal window, so that not all
    functions that use these values continually have to keep track of
    it separately.

    Only works if common_handle_winch is registered to handle winch signals.
*/
int common_get_width();
/**
   Returns the height of the terminal window, so that not all
   functions that use these values continually have to keep track of
   it separatly.

   Only works if common_handle_winch is registered to handle winch signals.
*/
int common_get_height();

/**
  Handle a window change event by looking up the new window size and
  saving it in an internal variable used by common_get_wisth and
  common_get_height().
*/
void common_handle_winch(int signal);

/**
   Write paragraph of output to the specified stringbuffer, and redo
   the linebreaks to fit the current screen.
*/
void write_screen(const wcstring &msg, wcstring &buff);

/**
   Tokenize the specified string into the specified wcstring_list_t.
   \param val the input string. The contents of this string is not changed.
   \param out the list in which to place the elements.
*/
void tokenize_variable_array(const wcstring &val, wcstring_list_t &out);

/**
   Make sure the specified direcotry exists. If needed, try to create
   it and any currently not existing parent directories..

   \return 0 if, at the time of function return the directory exists, -1 otherwise.
*/
int create_directory(const wcstring &d);

/**
   Print a short message about how to file a bug report to stderr
*/
void bugreport();

/**
   Return the number of seconds from the UNIX epoch, with subsecond
   precision. This function uses the gettimeofday function, and will
   have the same precision as that function.

   If an error occurs, NAN is returned.
 */
double timef();

/**
	Call the following function early in main to set the main thread.
    This is our replacement for pthread_main_np().
 */
void set_main_thread();
bool is_main_thread();

/** Configures thread assertions for testing */
void configure_thread_assertions_for_testing();

/** Set up a guard to complain if we try to do certain things (like take a lock) after calling fork */
void setup_fork_guards(void);

/** Save the value of tcgetpgrp so we can restore it on exit */
void save_term_foreground_process_group(void);
void restore_term_foreground_process_group(void);

/** Return whether we are the child of a fork */
bool is_forked_child(void);
void assert_is_not_forked_child(const char *who);
#define ASSERT_IS_NOT_FORKED_CHILD_TRAMPOLINE(x) assert_is_not_forked_child(x)
#define ASSERT_IS_NOT_FORKED_CHILD() ASSERT_IS_NOT_FORKED_CHILD_TRAMPOLINE(__FUNCTION__)

extern "C" {
    __attribute__((noinline)) void debug_thread_error(void);
}


#endif
