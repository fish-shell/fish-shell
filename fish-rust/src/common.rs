use crate::ffi;
use crate::wchar_ext::WExt;
use crate::wchar_ffi::c_str;
use crate::wchar_ffi::{wstr, WCharFromFFI, WString};
use std::{ffi::c_uint, mem};

/// A scoped manager to save the current value of some variable, and optionally set it to a new
/// value. When dropped, it restores the variable to its old value.
///
/// This can be handy when there are multiple code paths to exit a block.
pub struct ScopedPush<'a, T> {
    var: &'a mut T,
    saved_value: Option<T>,
}

impl<'a, T> ScopedPush<'a, T> {
    pub fn new(var: &'a mut T, new_value: T) -> Self {
        let saved_value = mem::replace(var, new_value);

        Self {
            var,
            saved_value: Some(saved_value),
        }
    }

    pub fn restore(&mut self) {
        if let Some(saved_value) = self.saved_value.take() {
            *self.var = saved_value;
        }
    }
}

impl<'a, T> Drop for ScopedPush<'a, T> {
    fn drop(&mut self) {
        self.restore()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EscapeStringStyle {
    Script(EscapeFlags),
    Url,
    Var,
    Regex,
}

/// Flags for the [`escape_string()`] function. These are only applicable when the escape style is
/// [`EscapeStringStyle::Script`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct EscapeFlags {
    /// Do not escape special fish syntax characters like the semicolon. Only escape non-printable
    /// characters and backslashes.
    pub no_printables: bool,
    /// Do not try to use 'simplified' quoted escapes, and do not use empty quotes as the empty
    /// string.
    pub no_quoted: bool,
    /// Do not escape tildes.
    pub no_tilde: bool,
    /// Replace non-printable control characters with Unicode symbols.
    pub symbolic: bool,
}

/// Replace special characters with backslash escape sequences. Newline is replaced with `\n`, etc.
pub fn escape_string(s: &wstr, style: EscapeStringStyle) -> WString {
    let mut flags_int = 0;

    let style = match style {
        EscapeStringStyle::Script(flags) => {
            const ESCAPE_NO_PRINTABLES: c_uint = 1 << 0;
            const ESCAPE_NO_QUOTED: c_uint = 1 << 1;
            const ESCAPE_NO_TILDE: c_uint = 1 << 2;
            const ESCAPE_SYMBOLIC: c_uint = 1 << 3;

            if flags.no_printables {
                flags_int |= ESCAPE_NO_PRINTABLES;
            }
            if flags.no_quoted {
                flags_int |= ESCAPE_NO_QUOTED;
            }
            if flags.no_tilde {
                flags_int |= ESCAPE_NO_TILDE;
            }
            if flags.symbolic {
                flags_int |= ESCAPE_SYMBOLIC;
            }

            ffi::escape_string_style_t::STRING_STYLE_SCRIPT
        }
        EscapeStringStyle::Url => ffi::escape_string_style_t::STRING_STYLE_URL,
        EscapeStringStyle::Var => ffi::escape_string_style_t::STRING_STYLE_VAR,
        EscapeStringStyle::Regex => ffi::escape_string_style_t::STRING_STYLE_REGEX,
    };

    ffi::escape_string(c_str!(s), flags_int.into(), style).from_ffi()
}

/// Test if the string is a valid function name.
pub fn valid_func_name(name: &wstr) -> bool {
    if name.is_empty() {
        return false;
    };
    if name.char_at(0) == '-' {
        return false;
    };
    // A function name needs to be a valid path, so no / and no NULL.
    if name.find_char('/').is_some() {
        return false;
    };
    if name.find_char('\0').is_some() {
        return false;
    };
    true
}

pub const fn assert_send<T: Send>() {}

pub const fn assert_sync<T: Sync>() {}
