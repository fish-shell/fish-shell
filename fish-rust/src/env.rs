/// Flags that may be passed as the 'mode' in env_stack_t::set() / environment_t::get().
pub mod flags {
    use autocxx::c_int;

    /// Default mode. Used with `env_stack_t::get()` to indicate the caller doesn't care what scope
    /// the var is in or whether it is exported or unexported.
    pub const ENV_DEFAULT: c_int = c_int(0);
    /// Flag for local (to the current block) variable.
    pub const ENV_LOCAL: c_int = c_int(1 << 0);
    pub const ENV_FUNCTION: c_int = c_int(1 << 1);
    /// Flag for global variable.
    pub const ENV_GLOBAL: c_int = c_int(1 << 2);
    /// Flag for universal variable.
    pub const ENV_UNIVERSAL: c_int = c_int(1 << 3);
    /// Flag for exported (to commands) variable.
    pub const ENV_EXPORT: c_int = c_int(1 << 4);
    /// Flag for unexported variable.
    pub const ENV_UNEXPORT: c_int = c_int(1 << 5);
    /// Flag to mark a variable as a path variable.
    pub const ENV_PATHVAR: c_int = c_int(1 << 6);
    /// Flag to unmark a variable as a path variable.
    pub const ENV_UNPATHVAR: c_int = c_int(1 << 7);
    /// Flag for variable update request from the user. All variable changes that are made directly
    /// by the user, such as those from the `read` and `set` builtin must have this flag set. It
    /// serves one purpose: to indicate that an error should be returned if the user is attempting
    /// to modify a var that should not be modified by direct user action; e.g., a read-only var.
    pub const ENV_USER: c_int = c_int(1 << 8);
}

/// Return values for `env_stack_t::set()`.
pub mod status {
    pub const ENV_OK: i32 = 0;
    pub const ENV_PERM: i32 = 1;
    pub const ENV_SCOPE: i32 = 2;
    pub const ENV_INVALID: i32 = 3;
    pub const ENV_NOT_FOUND: i32 = 4;
}
