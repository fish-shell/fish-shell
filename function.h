/** \file function.h

    Prototypes for functions for storing and retrieving function
  information. These functions also take care of autoloading
  functions in the $fish_function_path. Actual function evaluation
  is taken care of by the parser and to some degree the builtin
  handling library.
*/

#ifndef FISH_FUNCTION_H
#define FISH_FUNCTION_H

#include <wchar.h>
#include <map>

#include "util.h"
#include "common.h"
#include "event.h"
#include "env.h"

class parser_t;
class env_vars_snapshot_t;

/**
   Structure describing a function. This is used by the parser to
   store data on a function while parsing it. It is not used
   internally to store functions, the function_internal_data_t
   structure is used for that purpose. Parhaps these two should be
   merged.
  */
struct function_data_t
{
    /**
       Name of function
     */
    wcstring name;
    /**
       Description of function
     */
    wcstring description;
    /**
       Function definition
     */
    const wchar_t *definition;
    /**
       List of all event handlers for this function
     */
    std::vector<event_t> events;
    /**
       List of all named arguments for this function
     */
    wcstring_list_t named_arguments;
    /**
       List of all variables that are inherited from the function definition scope.
       The variable values are snapshotted when function_add() is called.
      */
    wcstring_list_t inherit_vars;
    /**
       Set to non-zero if invoking this function shadows the variables
       of the underlying function.
     */
    int shadows;
};

class function_info_t
{
public:
    /** Constructs relevant information from the function_data */
    function_info_t(const function_data_t &data, const wchar_t *filename, int def_offset, bool autoload);

    /** Used by function_copy */
    function_info_t(const function_info_t &data, const wchar_t *filename, int def_offset, bool autoload);

    /** Function definition */
    const wcstring definition;

    /** Function description. Only the description may be changed after the function is created. */
    wcstring description;

    /** File where this function was defined (intern'd string) */
    const wchar_t * const definition_file;

    /** Line where definition started */
    const int definition_offset;

    /** List of all named arguments for this function */
    const wcstring_list_t named_arguments;

    /** Mapping of all variables that were inherited from the function definition scope to their values */
    const std::map<wcstring,env_var_t> inherit_vars;

    /** Flag for specifying that this function was automatically loaded */
    const bool is_autoload;

    /** Set to true if invoking this function shadows the variables of the underlying function. */
    const bool shadows;
};


/**
   Initialize function data
*/
void function_init();

/** Add a function. definition_line_offset is the line number of the function's definition within its source file */
void function_add(const function_data_t &data, const parser_t &parser, int definition_line_offset = 0);

/**
   Remove the function with the specified name.
*/
void function_remove(const wcstring &name);

/**
   Returns by reference the definition of the function with the name \c name.
   Returns true if successful, false if no function with the given name exists.
*/
bool function_get_definition(const wcstring &name, wcstring *out_definition);

/**
   Returns by reference the description of the function with the name \c name.
   Returns true if the function exists and has a nonempty description, false if it does not.
*/
bool function_get_desc(const wcstring &name, wcstring *out_desc);

/**
   Sets the description of the function with the name \c name.
*/
void function_set_desc(const wcstring &name, const wcstring &desc);

/**
   Returns true if the function with the name name exists.
*/
int function_exists(const wcstring &name);

/**
   Returns true if the function with the name name exists, without triggering autoload.
*/
int function_exists_no_autoload(const wcstring &name, const env_vars_snapshot_t &vars);

/**
   Returns all function names.

   \param get_hidden whether to include hidden functions, i.e. ones starting with an underscore
*/
wcstring_list_t function_get_names(int get_hidden);

/**
   Returns tha absolute path of the file where the specified function
   was defined. Returns 0 if the file was defined on the commandline.

   This function does not autoload functions, it will only work on
   functions that have already been defined.

   This returns an intern'd string.
*/
const wchar_t *function_get_definition_file(const wcstring &name);

/**
   Returns the linenumber where the definition of the specified
   function started.

   This function does not autoload functions, it will only work on
   functions that have already been defined.
*/
int function_get_definition_offset(const wcstring &name);

/**
   Returns a list of all named arguments of the specified function.
*/
wcstring_list_t function_get_named_arguments(const wcstring &name);

/**
   Returns a mapping of all variables of the specified function that were inherited
   from the scope of the function definition to their values.
 */
std::map<wcstring,env_var_t> function_get_inherit_vars(const wcstring &name);

/**
   Creates a new function using the same definition as the specified function.
   Returns true if copy is successful.
*/
bool function_copy(const wcstring &name, const wcstring &new_name);


/**
   Returns whether this function shadows variables of the underlying function
*/
int function_get_shadows(const wcstring &name);

#endif
