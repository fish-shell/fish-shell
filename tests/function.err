function: function name shadows a builtin so you must use '--shadow-builtin'
fish: function pwd; end
      ^
yes, it failed as expected
function: function name does not shadow a builtin so you must not use '--shadow-builtin'
fish: function not_builtin --shadow-builtin; end
      ^
yes, it failed as expected
