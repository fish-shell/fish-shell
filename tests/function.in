# vim: set filetype=fish:
#
# Test the `function` builtin

# utility function
function show_ary -a name --no-scope-shadowing
    set -l count (count $$name)
    echo "\$$name: ($count)"
    if test $count -gt 0
        for i in (seq $count)
            echo "$i: '$$name[1][$i]'"
        end
    end
end

# Test the -V flag
set -g foo 'global foo'
set -l foo 'local foo'
set bar one 'two    2' \t '' 3
set baz
function frob -V foo -V bar -V baz
    show_ary foo
    show_ary bar
    show_ary baz
end
echo "Testing -V"
frob
echo "Testing -V with changed variables"
set foo 'bad foo'
set bar 'bad bar'
set baz 'bad baz'
frob

# Test that -a does not mix up the function name with arguments
# See #2068
function name1 -a arg1 arg2 ; end
function -a arg1 arg2 name2 ; end
function name3 --argument-names arg1 arg2 ; end
function --argument-names arg1 arg2 name4 ; end
for i in (seq 4)
    if functions -q name$i
        echo "Function name$i found"
    else
        echo "Function name$i not found, but should have been"
    end
end

# Test that we can't define a function that shadows a builtin by accident.
function pwd; end
or echo 'yes, it failed as expected' >&2

# Test that we can define a function that shadows a builtin if we use the
# right flag.
function pwd --shadow-builtin; end
and echo '"function pwd --shadow-builtin" worked'

# Using --shadow-builtin for a non-builtin function name also fails.
function not_builtin --shadow-builtin; end
or echo 'yes, it failed as expected' >&2
