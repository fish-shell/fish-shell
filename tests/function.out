Testing -V
$foo: (1)
1: 'local foo'
$bar: (5)
1: 'one'
2: 'two    2'
3: '	'
4: ''
5: '3'
$baz: (0)
Testing -V with changed variables
$foo: (1)
1: 'local foo'
$bar: (5)
1: 'one'
2: 'two    2'
3: '	'
4: ''
5: '3'
$baz: (0)
Function name1 found
Function name2 found
Function name3 found
Function name4 found
"function pwd --shadow-builtin" worked
